source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp

```diff
@@ -0,0 +1,637 @@
+        {
+          m_requests.erase(it);
+        }
+      }
+      void swap_request_to_signal(ReadRequest* original, ReadRequest* newRequest)
+      {
+        const rsl::unique_lock lock(m_requests_access_mtx);
+        auto it = rsl::find(m_requests.begin(), m_requests.end(), original);
+
+        REX_ASSERT_X(it != m_requests.end(), "Request at {} not found as a request to signal", reinterpret_cast<void*>(original)); // NOLINT(cppcoreguidelines-pro-type-reinterpret-cast)
+
+        *it = newRequest;
+      }
+
+      void signal_requests(const rsl::byte* buffer, rsl::memory_size size)
+      {
+        const rsl::unique_lock lock(m_requests_access_mtx);
+
+        for(ReadRequest* read_request : m_requests)
+        {
+          read_request->signal(buffer, size);
+        }
+      }
+
+      bool all_requests_finished() const
+      {
+        return m_requests.empty();
+      }
+
+      rsl::string_view filepath() const
+      {
+        return m_filepath;
+      }
+
+    private:
+      rsl::medium_stack_string m_filepath;
+      rsl::vector<ReadRequest*> m_requests;
+      rsl::mutex m_requests_access_mtx;
+      rsl::atomic<bool> m_is_done;
+    };
+
+    ReadRequest::ReadRequest(rsl::string_view filepath, QueuedRequest* queuedRequest)
+        : m_filepath(filepath)
+        , m_queued_request(queuedRequest)
+        , m_is_done(false)
+        , m_buffer(nullptr)
+        , m_size(0_bytes)
+    {
+    }
+
+    ReadRequest::ReadRequest(const ReadRequest& other)
+        : m_filepath(other.m_filepath)
+        , m_queued_request(other.m_queued_request)
+        , m_is_done(other.m_is_done)
+        , m_buffer(other.m_buffer)
+        , m_size(other.m_size)
+    {
+      m_queued_request->add_request_to_signal(this);
+    }
+
+    ReadRequest::ReadRequest(ReadRequest&& other)
+        : m_filepath(rsl::exchange(other.m_filepath, ""))
+        , m_queued_request(rsl::exchange(other.m_queued_request, nullptr))
+        , m_is_done(rsl::exchange(other.m_is_done, false))
+        , m_buffer(rsl::exchange(other.m_buffer, nullptr))
+        , m_size(rsl::exchange(other.m_size, 0_bytes))
+    {
+      m_queued_request->swap_request_to_signal(&other, this);
+    }
+
+    ReadRequest::~ReadRequest()
+    {
+      if(m_queued_request)
+      {
+        m_queued_request->remove_request_to_signal(this);
+      }
+    }
+
+    ReadRequest& ReadRequest::operator=(const ReadRequest& other) // NOLINT(bugprone-unhandled-self-assignment)
+    {
+      REX_ASSERT_X(this != &other, "assigning a read request to itself");
+
+      m_queued_request->remove_request_to_signal(this);
+
+      m_filepath       = other.m_filepath;
+      m_queued_request = other.m_queued_request;
+      m_is_done        = other.m_is_done;
+      m_buffer         = other.m_buffer;
+      m_size           = other.m_size;
+
+      m_queued_request->add_request_to_signal(this);
+
+      return *this;
+    }
+    ReadRequest& ReadRequest::operator=(ReadRequest&& other)
```

Comment:
not needed

Reply:
I add this keyword to all the "move" ctors or move assignment operators are you telling me this is redundant?
Is it because we don't have expections enabled?

Reply:
yes